compiler/rustc_codegen_ssa/src/mir/naked_asm.rs

```diff
@@ -0,0 +1,248 @@
+
+            GlobalAsmOperandRef::SymFn { instance }
+        }
+        InlineAsmOperand::SymStatic { def_id } => {
+            GlobalAsmOperandRef::SymStatic { def_id: *def_id }
+        }
+        InlineAsmOperand::In { .. }
+        | InlineAsmOperand::Out { .. }
+        | InlineAsmOperand::InOut { .. }
+        | InlineAsmOperand::Label { .. } => {
+            bug!("invalid operand type for naked_asm!")
+        }
+    }
+}
+
+enum AsmBinaryFormat {
+    Elf,
+    Macho,
+    Coff,
+}
+
+impl AsmBinaryFormat {
+    fn from_target(target: &rustc_target::spec::Target) -> Self {
+        if target.is_like_windows {
+            Self::Coff
+        } else if target.is_like_osx {
+            Self::Macho
+        } else {
+            Self::Elf
+        }
+    }
+}
+
+fn prefix_and_suffix<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    instance: Instance<'tcx>,
+    asm_name: &str,
+    item_data: &MonoItemData,
+) -> (String, String) {
+    use std::fmt::Write;
+
+    let is_arm = tcx.sess.target.arch == "arm";
+    let is_thumb = tcx.sess.unstable_target_features.contains(&sym::thumb_mode);
+
+    let attrs = tcx.codegen_fn_attrs(instance.def_id());
+    let link_section = attrs.link_section.map(|symbol| symbol.as_str().to_string());
+    let align = attrs.alignment.map(|a| a.bytes()).unwrap_or(4);
+
+    // See https://sourceware.org/binutils/docs/as/ARM-Directives.html for info on these directives.
+    // In particular, `.arm` can also be written `.code 32` and `.thumb` as `.code 16`.
+    let (arch_prefix, arch_suffix) = if is_arm {
+        (
+            match attrs.instruction_set {
+                None => match is_thumb {
+                    true => ".thumb\n.thumb_func",
+                    false => ".arm",
+                },
+                Some(InstructionSetAttr::ArmT32) => ".thumb\n.thumb_func",
+                Some(InstructionSetAttr::ArmA32) => ".arm",
+            },
+            match is_thumb {
+                true => ".thumb",
+                false => ".arm",
+            },
+        )
+    } else {
+        ("", "")
+    };
+
+    // see https://godbolt.org/z/cPK4sxKor
+    let linkage = match item_data.linkage {
+        Linkage::External => ".globl",
+        Linkage::LinkOnceAny => ".weak",
+        Linkage::LinkOnceODR => ".weak",
+        Linkage::WeakAny => ".weak",
+        Linkage::WeakODR => ".weak",
+        Linkage::Internal => ".local",
+        Linkage::Private => ".local",
+        Linkage::AvailableExternally => {
+            // this would make the function equal an extern definition
+            bug!("Functions may not have available_externally linkage")
+        }
+        Linkage::Appending => bug!("Only global variables can have appending linkage!"),
+        Linkage::Common => bug!("Functions may not have common linkage"),
+        Linkage::ExternalWeak => {
+            // FIXME: actually this causes a SIGILL in LLVM
+            bug!("Functions may not have external weak linkage")
+        }
+    };
```

Comment:
@bjorn3 that worked out well, thanks for that hint!

It does reintroduce this match. I plan to make a separate PR that limits the usage of the linkage attribute so that at least e.g. common linkage on a function gets rejected nicely earlier on.

Reply:
For the cases where we don't yet check that it isn't used, I think emitting a fatal error rather than a `bug!()` would be nicer, but then again `#[linkage]` can be abused in many ways already (some likely causing an ICE or LLVM crash), so it isn't all that important to avoid an ICE here IMO.